### ISSUE TITLE
Memory leak in k2.ragged.normalize_scores

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#706]
https://github.com/k2-fsa/k2/blob/36d188b23844121a78f95203ba470708be691645/k2/python/k2/ragged/autograd.py#L65

There is a memory leak when we do:
```
  ctx.out = out[0]  # save for backward      
```
When you do `ctx.save_for_backward(something)`, it does some magic stuff to prevent reference cycles. 
The problem is that if we have requires_grad=True, `out[0].values` will end up having a reference to
the `ctx` object indirectly held via its grad_fn; and saving ctx.out without doing .detach() will cause a reference
cycle.

I think the correct fix for this is to add a .detach() funtion to class RaggedFloat, and do
`ctx.out = out[0].detach()`.

**Comment by danpovey** [comment]
.. i discovered this by noticing, when debugging possible memory leaks, that the P_scores that we create on
every iteration appears to never be freed.  (This will eventually use up quite a bit of GPU memory, at least on
very long runs).

**Comment by csukuangfj** [comment]
Yes, you are right. There are memory leaks.

I just used the following changes and example code to verify it.

## Changes

```diff
diff --git a/k2/python/k2/ragged/tensor.py b/k2/python/k2/ragged/tensor.py
index 3a1cb85..b10d779 100644
--- a/k2/python/k2/ragged/tensor.py
+++ b/k2/python/k2/ragged/tensor.py
@@ -20,6 +20,9 @@ class RaggedFloat(object):
     Currently, it is used only in `k2.ragged.normalize_scores`.
     '''

+    def __del__(self):
+        print(f'del {id(self)}')
+
     def __init__(self,
                  ragged: Union[str, _k2.RaggedFloat, _k2.RaggedShape],
                  values: Optional[torch.Tensor] = None):
@@ -42,6 +45,7 @@ class RaggedFloat(object):
             :class:`_k2.RaggedShape`. It is a 1-D torch.Tensor with dtype
             torch.float32.
         '''
+        print(f'init {id(self)}')
         if isinstance(ragged, str):
             ragged = _k2.RaggedFloat(ragged)
             assert values is None
```

## Example code
```python
#!/usr/bin/env python3

import k2


def main():
    a = k2.RaggedFloat('[[1.]]')
    b = k2.ragged.RaggedFloat(a)
    b.requires_grad_(True)
    c = k2.ragged.normalize_scores(b, True)
    print(c)


if __name__ == '__main__':
    main()
```

### Output
```
init 139839467423056
init 139839467421856
[ [ 0 ] ]
del 139839467423056
```

You can see that one object is not freed as its destructor is not called.

## Fix

```diff
     @staticmethod
     def forward(ctx, src: RaggedFloat, use_log: bool, out: List[RaggedFloat],
                 unused_scores: torch.Tensor) -> torch.Tensor:
@@ -63,17 +62,17 @@ class _NormalizeScores(torch.autograd.Function):
         assert len(out) == 1
         ans_ragged = _k2.normalize_per_sublist(src.ragged, use_log)
         out[0] = RaggedFloat(ans_ragged)
-        ctx.out = out[0]  # save for backward
+        ctx.out_ragged = out[0].ragged  # save for backward
         ctx.use_log = use_log
         return out[0].values

     @staticmethod
     def backward(ctx,
                  out_grad: torch.Tensor) -> Tuple[None, None, torch.Tensor]:
-        out = ctx.out
+        out_ragged = ctx.out_ragged
         use_log = ctx.use_log
         assert use_log is True
-        ans_grad = _k2.normalize_per_sublist_backward(out.ragged, use_log,
+        ans_grad = _k2.normalize_per_sublist_backward(out_ragged, use_log,
                                                       out_grad)

         return (
```

### Output after the fix

```
init 140265888429392
init 140265888428192
[ [ 0 ] ]
del 140265888429392
del 140265888428192
```

**Comment by danpovey** [comment]
Great-- thanks!!
